change code so setting up velocitiy constraints is faster, error if they are nonlinear

Peter230655 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.mechanics
  * KanesMethod requires linear speed constraints, but it does not check whether they indeed are linear. Using _linear_eq_to_matrix_ instead of the Jacobian serves two purposes: 1) it makes solving the constraints faster and 2) it will raise an error in the speed constraints are not linear.

<!-- END RELEASE NOTES -->

Peter230655 committed:
change code so KanesMethod will find nonlinearities in speed constraints

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -366,10 +366,10 @@ def _initialize_kindiffeq_matrices(self, kdeqs, linear_solver='LU'):
+            # so check for u in the components.
             dy_syms = find_dynamicsymbols(k_ku.row_join(k_kqdot).row_join(f_k))
-            nonlin_vars = [vari for vari in u[:] + qdot[:] if vari in dy_syms]
+            nonlin_vars = [vari for vari in u[:] if vari in dy_syms]
```
The changes shouldn't be made to the kinematical differential equations, those do have both u's and qdot's.

Peter230655 replied:
> The changes shouldn't be made to the kinematical differential equations, those do have both u's and qdot's.

oh, sh.....!!

Peter230655 committed:
mix up lien vel constr / differential equations

Peter230655 committed:
Merge branch 'sympy:master' into KanesMethodComplex_1

smichr commented:
A test that catches this error should be added.

Peter230655 commented:
I checked the speeds with my 40 link closed pendulum, as suggested:
- checking for linearity of the kinematic DEs: 0.0 sec
- checking for linearity of the velocity constraints: 0.0sec to 0.01sec

It takes about 35 sec to set up KanesEquations in this example.
I tested like this:
´´´
zeit = time.time()
.......
code to be checked
......
duration = time.time() - zeit
```

Peter230655 commented:
> A test that catches this error should be added.

You added the label _PR: author's turn_.
I am unclear what changes I should submit.
Thanks!

Peter230655 committed:
no changes

moorepants commented:
Doesn't the 40 link pendulum not have any speed constraints? What are you checking in that case?

Peter230655 commented:
I used the version where the other end is also attached to a point. Sorry, should hsve mentione it.

moorepants commented:
>  I am unclear what changes I should submit.

@smichr wrote "A test that catches this error should be added."

Peter230655 commented:
> > I am unclear what changes I should submit.
> 
> @smichr wrote "A test that catches this error should be added."

I did make a test, it is PR # 26901, maybe this is why I did not understand what was meant.
Is this o.k.?

moorepants commented:
The test needs to be in the same PR as the change, i.e. this PR.

Peter230655 committed:
moved the test to the main PR

Peter230655 committed:
Merge branch 'KanesMethodComplex_1' of github.com:Peter230655/sympy-pst into KanesMethodComplex_1

Peter230655 commented:
I added the test for nonlinearity of velocity constraints to this PR, I will close the other PR, which only held the test

Peter230655 commented:
I checked the time needed to check the velocity constrains for nonlinearity again, using an old simulation of mine. Basically the skateboard of your lecture, modified a bit.
It has 6 velocity constraints.
Checking for nonlinearity took between 0.0 sec and 0.002 sec, total time to get kanes eoms took 1.3 sec.
I used the same method mentioned abvoe to check for the time:
```
zeit = time.time()
code to check for nonlinearity
.....
....
time_needed = time.tim() - zeit
```

tjstienstra commented on sympy/physics/mechanics/kane.py:
```diff
@@ -317,6 +317,17 @@ def _initialize_constraint_matrices(self, config, vel, acc, linear_solver='LU'):
 
+            # The linear velocity constraints must be linear in u, so check'
+            # for u in the components.
+            dy_syms = find_dynamicsymbols(self._f_nh.row_join(self._k_nh))
```
If you use [linear_eq_to_matrix](https://docs.sympy.org/latest/modules/solvers/solveset.html#sympy.solvers.solveset.linear_eq_to_matrix) you'll speed up the creation of the matrices, while raising a NonlinearError if the velocity constraints are nonlinear. E.g.
```python
self._k_nh, self._f_nh = linear_eq_to_matrix(vel, self.u)
```

Peter230655 replied:
Thanks!
But would this not, in logical consequence, entail rewriting _def initialize_constraint_matrices_ completely?

Peter230655 replied:
Thanks!
But would this not, in logical consequence, entail rewriting _def initialize_constraint_matrices_ completely?

tjstienstra replied:
> But would this not, in logical consequence, entail rewriting def initialize_constraint_matrices(.....) completely?

It would entail replacing the lines computing the matrix and vector:
```python
self._k_nh = ...jacobian(...)
self._f_nh = ...
```
While the rest of the function can remain unchanged.

The `NonlinearError` would indeed be raised by `linear_eq_to_matrix` with `initialize_constraint_matrices` in the stack trace, i.e. you can also see that it was raised when calling `linear_eq_to_matrix` here. If this error is not clear enough, then an option would be to catch the NonlinearError and update the message. Something like:
```python
try:
    ... = linear_eq_to_matrix(...)
except NonlinearError as e:
    msg = "Some clear message."
    raise NonlinearError(msg) from e
```
However, I would expect that the error message given by `linear_eq_to_matrix` is clear enough.

Peter230655 replied:
I tried as you suggested:
```
            if self._qdot_u_map is not None:
                vel = msubs(vel, self._qdot_u_map)
            self._f_nh, self._k_nh = linear_eq_to_matrix([vel, self.u])
        #    self._f_nh = msubs(vel, u_zero)
        #    self._k_nh = (vel - self._f_nh).jacobian(self.u)
```
but got this error message:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[9], [line 8](vscode-notebook-cell:?execution_count=9&line=8)
      [6](vscode-notebook-cell:?execution_count=9&line=6) speed_constraint = [i - j for i, j in zip(loesung, u_dep)]
      [7](vscode-notebook-cell:?execution_count=9&line=7) print(len(speed_constraint))
----> [8](vscode-notebook-cell:?execution_count=9&line=8) KM = KanesMethodComplex(N, q_ind=q_ind, u_ind=u_ind, u_dependent=u_dep, kd_eqs=kd, velocity_constraints=speed_constraint)
      [9](vscode-notebook-cell:?execution_count=9&line=9) (fr, frstar) = KM.kanes_equations(BODY, FL)
     [11](vscode-notebook-cell:?execution_count=9&line=11) MM = KM.mass_matrix_full

Cell In[2], [line 235](vscode-notebook-cell:?execution_count=2&line=235)
    [233](vscode-notebook-cell:?execution_count=2&line=233) _validate_coordinates(self.q, self.u)
    [234](vscode-notebook-cell:?execution_count=2&line=234) self._initialize_kindiffeq_matrices(kd_eqs, kd_eqs_solver)
--> [235](vscode-notebook-cell:?execution_count=2&line=235) self._initialize_constraint_matrices(
    [236](vscode-notebook-cell:?execution_count=2&line=236)     configuration_constraints, velocity_constraints,
    [237](vscode-notebook-cell:?execution_count=2&line=237)     acceleration_constraints, constraint_solver)

Cell In[2], [line 305](vscode-notebook-cell:?execution_count=2&line=305)
    [303](vscode-notebook-cell:?execution_count=2&line=303)     if self._qdot_u_map is not None:
    [304](vscode-notebook-cell:?execution_count=2&line=304)         vel = msubs(vel, self._qdot_u_map)
--> [305](vscode-notebook-cell:?execution_count=2&line=305)     self._f_nh, self._k_nh = linear_eq_to_matrix([vel, self.u])
    [306](vscode-notebook-cell:?execution_count=2&line=306) #    self._f_nh = msubs(vel, u_zero)
    [307](vscode-notebook-cell:?execution_count=2&line=307) #    self._k_nh = (vel - self._f_nh).jacobian(self.u)
    [308](vscode-notebook-cell:?execution_count=2&line=308)     # If no acceleration constraints given, calculate them.
    [309](vscode-notebook-cell:?execution_count=2&line=309)     if not acc:

File c:\Users\Peter\anaconda3\envs\sympy-dev\Lib\site-packages\sympy\solvers\solveset.py:2845, in linear_eq_to_matrix(equations, *symbols)
   [2767](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2767) r"""
   [2768](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2768) Converts a given System of Equations into Matrix form.
   [2769](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2769) Here `equations` must be a linear system of equations in
   (...)
   [2842](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2842)     be done *before* calling this routine.
   [2843](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2843) """
   [2844](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2844) if not symbols:
-> [2845](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2845)     raise ValueError(filldedent('''
   [2846](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2846)         Symbols must be given, for which coefficients
   [2847](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2847)         are to be found.
   [2848](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2848)         '''))
   [2850](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2850) # Check if 'symbols' is a set and raise an error if it is
   [2851](file:///C:/Users/Peter/anaconda3/envs/sympy-dev/Lib/site-packages/sympy/solvers/solveset.py:2851) if isinstance(symbols[0], set):

ValueError: 
Symbols must be given, for which coefficients are to be found.
```

moorepants replied:
It doesn't look like you are using the function properly. See the documentation:

https://docs.sympy.org/latest/modules/solvers/solveset.html#sympy.solvers.solveset.linear_eq_to_matrix

Peter230655 replied:
> It doesn't look like you are using the function properly. See the documentation:
> 
> https://docs.sympy.org/latest/modules/solvers/solveset.html#sympy.solvers.solveset.linear_eq_to_matrix

Now I see, I simply copied Timo, without looking.

Peter230655 commented:
I managed the _linear_eq_to_matrix_ part. But it seems, that the method must be change at other places.
On the other had, it seems to cut the time needed by around 50% from 0.36 sec to 0.17sec.
Worth pursuing?
If yes, I guess this would be a new PR, as the idea has changed: speed up calculating the velocity constraints.

moorepants commented:
It doesn't need a new pr, but you can change the title to reflect what it has become.

Peter230655 commented:
> It doesn't need a new pr, but you can change the title to reflect what it has become.

I will do so, if I get it going. :-)

Peter230655 commented:
It seems to work, and definitely 50% to 80% faster, but I am running in this problem:
linear_eq_to_matrix raises a PolyNonlinearError and a NonlinearError, but neither one seems 'defined': When I write
```
try:
    code
    ......
except NonlinearError as e:
    raise NonlineaeError('message')
```
I get the message NonlinearError is not defined or similar.
Same with 
```
with pytest.raises(NonlinearError):
```

tjstienstra commented:
> I get the message NonlinearError is not defined or similar.

You have to correctly import NonlinearError:
```python
import sympy as sm
from sympy.solvers.solveset import NonlinearError

x = sm.Matrix(sm.symbols("x1:3"))
A = sm.Matrix([[1, 2], [3, 4]])
b = sm.Matrix([5, 6])
sm.linear_eq_to_matrix(A * x - b, x[:])  # Works fine

try:
    sm.linear_eq_to_matrix(A * x * x[0] - b, x[:])
except NonlinearError as e:
    msg = "Some custom message."
    print("here")
    raise NonlinearError(msg) from e
```
This raises the following error:
```bash
Traceback (most recent call last):
  File "...\sympy\solvers\solveset.py", line 2874, in linear_eq_to_matrix
    eq, c = _linear_eq_to_dict(equations, symbols)
  File "...\sympy\polys\matrices\linsolve.py", line 171, in _linear_eq_to_dict
    c, d = _lin_eq2dict(e, symset)
  File "...\sympy\polys\matrices\linsolve.py", line 199, in _lin_eq2dict
    ci, ti = _lin_eq2dict(ai, symset)
  File "...\sympy\polys\matrices\linsolve.py", line 219, in _lin_eq2dict
    raise PolyNonlinearError(filldedent('''
sympy.polys.solvers.PolyNonlinearError:
nonlinear cross-term: x1*(x1 + 2*x2)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "...\temp.py", line 10, in <module>
    sm.linear_eq_to_matrix(A * x * x[0] - b, x[:])
  File "...\sympy\solvers\solveset.py", line 2876, in linear_eq_to_matrix
    raise NonlinearError(str(err))
sympy.solvers.solveset.NonlinearError:
nonlinear cross-term: x1*(x1 + 2*x2)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "...\temp.py", line 14, in <module>
    raise NonlinearError(msg) from e
sympy.solvers.solveset.NonlinearError: Some custom message.
```

Peter230655 commented:
> > I get the message NonlinearError is not defined or similar.
> 
> Did you forget to import `NonlinearError`?
> 
> ```python
> import sympy as sm
> from sympy.solvers.solveset import NonlinearError
> 
> x = sm.Matrix(sm.symbols("x1:3"))
> A = sm.Matrix([[1, 2], [3, 4]])
> b = sm.Matrix([5, 6])
> sm.linear_eq_to_matrix(A * x - b, x[:])  # Works fine
> 
> try:
>     sm.linear_eq_to_matrix(A * x * x[0] - b, x[:])
> except NonlinearError as e:
>     msg = "Some custom message."
>     print("here")
>     raise NonlinearError(msg) from e
> ```
> 
> This raises the following error:
> 
> ```shell
> Traceback (most recent call last):
>   File "...\sympy\solvers\solveset.py", line 2874, in linear_eq_to_matrix
>     eq, c = _linear_eq_to_dict(equations, symbols)
>   File "...\sympy\polys\matrices\linsolve.py", line 171, in _linear_eq_to_dict
>     c, d = _lin_eq2dict(e, symset)
>   File "...\sympy\polys\matrices\linsolve.py", line 199, in _lin_eq2dict
>     ci, ti = _lin_eq2dict(ai, symset)
>   File "...\sympy\polys\matrices\linsolve.py", line 219, in _lin_eq2dict
>     raise PolyNonlinearError(filldedent('''
> sympy.polys.solvers.PolyNonlinearError:
> nonlinear cross-term: x1*(x1 + 2*x2)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "...\temp.py", line 10, in <module>
>     sm.linear_eq_to_matrix(A * x * x[0] - b, x[:])
>   File "...\sympy\solvers\solveset.py", line 2876, in linear_eq_to_matrix
>     raise NonlinearError(str(err))
> sympy.solvers.solveset.NonlinearError:
> nonlinear cross-term: x1*(x1 + 2*x2)
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "...\temp.py", line 14, in <module>
>     raise NonlinearError(msg) from e
> sympy.solvers.solveset.NonlinearError: Some custom message.
> ```

I did not forget - I did not know! :-)
Same with PolyNonlinearError?
Thanks!!

tjstienstra commented:
> Same with PolyNonlinearError?

```python
from sympy.polys.solvers import PolyNonlinearError
```
I wouldn't expect that you get a `PolyNonlinearError` as I guess that `linear_eq_to_matrix` actually catches those and changes them to `NonlinearError`s.

Peter230655 commented:
> > Same with PolyNonlinearError?
> 
> ```python
> from sympy.polys.solvers import PolyNonlinearError
> ```
> 
> I wouldn't expect that you get a `PolyNonlinearError` as I guess that `linear_eq_to_matrix` actually catches those and changes them to `NonlinearError`s.

I seem to get both:
```
PolyNonlinearError: nonlinear term: ux(t)**2

During handling of the above exception, another exception occurred:

NonlinearError                            Traceback (most recent call last)
File Untitled-1:306
    [305](untitled-1:305)     u_test = list(self.u)
--> [306](untitled-1:306)     self._k_nh, self._f_nh = linear_eq_to_matrix(vel_test, u_test)
    [308](untitled-1:308) # If no acceleration constraints given, calculate them.
```

tjstienstra commented:
> I seem to get both:

No, what happens is that `PolyNonlinearError` is catched using try except and a `NonlinearError` is raised, which is what you get. However, to give users more insight Python also shows the original error, i.e. the `PolyNonlinearError`, in the traceback.

Peter230655 commented:
> > I seem to get both:
> 
> No, what happens is that `PolyNonlinearError` is catched using try except and a `NonlinearError` is raised, which is what you get. However, to give users more insight Python also shows the original error, i.e. the `PolyNonlinearError`, in the traceback.

Thanks!
So should I skip the try.....except(NonlinearError) part, as you suggested originally?

tjstienstra commented:
> So should I skip the try.....except(NonlinearError) part, as you suggested originally?

No, I would just use `self._k_nh, self._f_nh = linear_eq_to_matrix(...)` directly without wrapping it in a if statement. The error raised by `linear_eq_to_matrix` is already quite clear.

Peter230655 commented:
Finally I got it to work, even with the try.....except!
**Thanks al lot - I did learn something today!!**

The increase in speed is leally substantial as per my inaccurate measurements!

Peter230655 committed:
speed up calculation of velocity constraints, raise error if nonlinear

Peter230655 commented:
I used Timo's idea of _linear_eq_to_matrix_ function to speed things up considerably. Raise error if nonlinear.

Peter230655 committed:
added blank line to test....

Peter230655 committed:
corrected some errors

tjstienstra commented on sympy/physics/mechanics/kane.py:
```diff
@@ -290,33 +292,35 @@ def _initialize_constraint_matrices(self, config, vel, acc, linear_solver='LU'):
-            if not acc:
-                _f_dnh = (self._k_nh.diff(dynamicsymbols._t) * self.u +
-                          self._f_nh.diff(dynamicsymbols._t))
+            try:
```
Good to see that you've figured out how to use the try except statement. However, we don't really require it here as the error is already clear enough. You can vastly simplify all changes to:
```python
    def _initialize_constraint_matrices(self, config, vel, acc, linear_solver='LU'):
        """Initializes constraint matrices."""
        ...
        if vel:
            # When calling kanes_equations, another class instance will be
            # created if auxiliary u's are present. In this case, the
            # computation of kinetic differential equation matrices will be
            # skipped as this was computed during the original KanesMethod
            # object, and the qd_u_map will not be available.
            if self._qdot_u_map is not None:
                vel = msubs(vel, self._qdot_u_map)

            self._k_nh, self._f_nh = linear_eq_to_matrix(vel, self.u[:])
            # If no acceleration constraints given, calculate them.
            if not acc:
                ...
            else:
                if self._qdot_u_map is not None:
                    acc = msubs(acc, self._qdot_u_map)
                self._k_dnh, self._f_dnh = linear_eq_to_matrix(acc, self._udot[:])

            ...
        else:
            ...
```

Peter230655 replied:
I am getting all sorts of errors! :-(
You think, they are related to my try.....except?

tjstienstra replied:
Hmmm, we are apparently defining things as `self._k_nh * self.u + self._f_nh == vel`, while `linear_eq_to_matrix` uses the form `Ax - b == eqs`.

P.S. I just updated my above code suggestion.

Peter230655 replied:
> Hmmm, we are apparently defining things as `self._k_nh * self.u + self._f_nh == vel`, while `linear_eq_to_matrix` uses the form `Ax - b == eqs` (if I'm not mistaken).
> 
> P.S. I've just updated my above code suggestion.

I have tried this, but still so many mistakes. :-((
But it is so much faster it has to work!

tjstienstra replied:
I'm having no problems. Here is a [patch](https://github.com/user-attachments/files/17210440/peter.patch) (the patch is w.r.t. master). You can either just view the contents of the file or checkout and apply it to master.

Peter230655 committed:
more errors corrected

Peter230655 committed:
Timo's version copied

Peter230655 commented:
I just copied Timo's version, hopefully correctly so.

Peter230655 committed:
removed unneeded import

Peter230655 commented:
I copied correctly! :-)
To be very clear: That this finaly worked was **ONLY** Timo's skill!

moorepants commented:
> KanesMethod allows for linear speed constraints, however it does not check whether they really are linear. It does test the kinematic differential equations for linearity. So, I simply copied the few lines of code there and added them around line 320

needs to be improved. State what is changes: "KanesMethod will now raise a NonLinearError if nonlinear velocity constraints are passed in."

Peter230655 commented:
> > KanesMethod allows for linear speed constraints, however it does not check whether they really are linear. It does test the kinematic differential equations for linearity. So, I simply copied the few lines of code there and added them around line 320
> 
> needs to be improved. State what is changes: "KanesMethod will now raise a NonLinearError if nonlinear velocity constraints are passed in."

I changed. It was NOT my idea, but Timo's. Can I mention it there?

moorepants commented:
> Can I mention it there?

No, that's not the place for it.

moorepants commented:
LGTM! Thanks.

Peter230655 commented:
> > Can I mention it there?
> 
> No, that's not the place for it.

That's what I expected :-) . I think, I made it clear, that it was not my merit.
Thanks for merging!! :-))
